modalsheet/src/main/java/eu/wewox/modalsheet/ModalSheet.kt

```diff
@@ -95,260 +84,158 @@ public fun <T> ModalSheet(
+            // Intercept and disallow hide gesture / action
+            if (it == ModalBottomSheetValue.Hidden && !cancelable) {
+                return@rememberModalBottomSheetState false
+            }
+
+            onVisibleChange(it == ModalBottomSheetValue.Expanded)
+
+            true
+        }
+    )
 
-    if (visible || contentVisible) {
-        FullscreenPopup(
-            onDismiss = onDismiss
-        ) {
-            PopupBody(
-                visible = visible,
-                scrim = scrim,
-                content = {
-                    // Set contentVisible true when the composable enters composition and false when it leaves
-                    DisposableEffect(Unit) {
-                        contentVisible = true
-                        onDispose { contentVisible = false }
-                    }
+    if (!visible && sheetState.progress.to == sheetState.progress.from && !sheetState.isVisible) {
+        return
+    }
 
-                    val scope = remember { ModalSheetScopeImpl() }
-                    SwipeableContent(
-                        swipeEnabled = swipeEnabled,
-                        onSwipedAway = onDismiss,
-                        scrollState = scope.scrollState
-                    ) {
-                        Surface(
-                            shape = shape,
-                            color = backgroundColor,
-                            modifier = Modifier.fillMaxWidth()
-                        ) {
-                            scope.content()
-                        }
-                    }
-                }
-            )
+    LaunchedEffect(visible) {
+        if (visible) {
+            sheetState.show()
+        } else {
+            sheetState.hide()
         }
     }
-}
-
-/**
- * The scope for [ModalSheet] content.
- */
-@ExperimentalSheetApi
-public interface ModalSheetScope {
 
-    /**
-     * The scroll state.
-     * Must be passed to [.verticalScroll] modifier to support scrolling inside content.
-     */
-    public val scrollState: ScrollState
+    ModalSheet(
+        sheetState = sheetState,
+        onDismiss = {
+            if (cancelable) {
+                onVisibleChange(false)
+            }
+        },
+        shape = shape,
+        elevation = elevation,
+        backgroundColor = backgroundColor,
+        contentColor = contentColor,
+        scrimColor = scrimColor,
+        content = content,
+    )
 }
 
 /**
- * Contains the default values used by [ModalSheet].
+ * Static modal sheet that behaves like bottom sheet and draws over system UI.
+ * Receives [ModalBottomSheetState] to fine-tune sheet behavior.
+ *
+ * Note: In this case [ModalSheet] is always added to the composition. See [ModalSheet] which receives visible flag, or
+ * data object to conditionally add / remove modal sheet to / from the composition.
+ *
+ * @param sheetState The state of the bottom sheet.
+ * @param onDismiss Called when user taps on the HW back button.
```

Comment:
I would consider not using technical abbreviations in documentation like "HW" (even though they are well known)